File: C#/49.group-anagrams.cs
/*
 * @lc app=leetcode id=49 lang=csharp
 *
 * [49] Group Anagrams
 */

// @lc code=start
public class Solution
{
    public IList<IList<string>> GroupAnagrams(string[] strs)
    {
        Dictionary<string, List<string>> map = new Dictionary<string, List<string>>();

        foreach (string str in strs)
        {
            string key = string.Join("", str.OrderBy(c => c));
            if (!map.ContainsKey(key))
            {
                map[key] = new List<string>();
            }

            map[key].Add(str);
        }

        return map.Values.ToList<IList<string>>();
    }
}
// @lc code=end

File: C++/array-and-hashing/49.group-anagrams.cpp
/*
 * @lc app=leetcode id=49 lang=cpp
 *
 * [49] Group Anagrams
 */

// @lc code=start
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map <string, vector<string>> map;

        for (auto str : strs)
        {
            string s = str;
            sort(s.begin(), s.end());

            map[s].push_back(str);
        }

        vector<vector<string>> result;

        for (auto pair : map) {
            result.push_back(pair.second);
        }

        return result;
    }
};
// @lc code=end

